docs/features/plugin/tools/getting-started.md

```diff
@@ -0,0 +1,270 @@
+
+            weather_description = data["weather"][0]["description"]
+            temperature = data["main"]["temp"]
+            humidity = data["main"]["humidity"]
+            wind_speed = data["wind"]["speed"]
+
+            return f"Weather in {city}: {temperature}°C"
+        except requests.RequestException as e:
+            return f"Error fetching weather data: {str(e)}"
+
+```
+---
+## Parameters
+To use parameters, you need to use type hints for all parameters directly in the function definition.
+:::tip
+ Reserved params do not require type hints.
+:::
+### Some of the supported types for tool parameters: String, Integer, Number, Boolean, Array, Object, Null, Literal, Optional.
+```python
+ async def get_bbc_news_content(
+        self,
+        uri: str,
+        __event_emitter__,
+        __user__,
+    ) -> str:
+    
+```
+> **Note:** Set default values for parameters to ensure the tool works properly, even when those parameters are not provided by the end user.
+
+
+
+## Valves and User Valves
+Valves are optional configurations that control a tool's behavior. The key difference is that valves can only be configured by admins, whereas user valves can be set by any user to suit their individual needs.
+
+They can be defined as follows:
+
+```python
+# Define and Valves
+    class Valves(BaseModel):
+        priority: int = Field(
+            default=0, description="Priority level for the filter operations."
+        )
+        test_valve: int = Field(
+            default=4, description="A valve controlling a numberical value"
+        )
+        pass
+
+    # Define any UserValves
+    class UserValves(BaseModel):
+        test_user_valve: bool = Field(
+            default=False, description="A user valve controlling a True/False (on/off) switch"
+        )
+        pass
+
+    def __init__(self):
+        self.valves = self.Valves()
+        pass
+```
+> **Note:** As you can see, you can give default value to valves and user valves.
+## Event Emitters
+Event emitters provide a simple and effective way to communicate messages, status updates, and errors to users during the execution of your tool. By including event emitter calls in your function, you can ensure that important updates are displayed in real-time while the final response is being generated.
+
+![Event_emitter Demo](/img/demo-event-emit.gif)
+
+You can pass event emitter directly to your function as input, and use it as follow:
+```python
+async def test_function(
+        self, prompt: str, __user__: dict, __event_emitter__=None
+    ) -> str:
+        """
+        This is a demo
+
+        :param test: this is a test parameter
+        """
+
+        await __event_emitter__(
+            {
+                "type": "status", # We set the type here
+                "data": {"description": "Message that shows up in the chat", "done": False}, 
+                # Note done is False here indicating we are still emitting statuses
```

Comment:
misleading, `"done": False` is used to display loading indicator from the frontend.